### PR TITLE
risc-v: reorganise traps.S

### PR DESCRIPTION
**Pull request by oliver-wm**
Move syscall and fastpath checks after interrupt and exceptions.
Exceptions were being interpreted as null-syscalls.
Remove duplication for return address.

Signed-off-by: Oliver Scott <Oliver.Scott@data61.csiro.au>

### CODE REVIEW

**Comment by oliver-wm** [comment]
This fixes the failing simulation and hardware builds on sel4test

File: src/arch/riscv/traps.S [hunk 1]
```diff
@@ -98,50 +98,43 @@ trap_entry:
...
-
-  li t3, SYSCALL_CALL
-  beq a7, t3, fp_call
-#endif
-
   /* Check if it's an interrupt */
   bltz s0, interrupt
 
-  li   s4, 8       /* ratified priv has value 8 for ecall exception */
+  /* ratified priv has value 8 for ecall exception */
```
Comment (axel-h) [inline src/arch/riscv/traps.S:104]: maybe be more precise and say: "for 'ecall from U-mode' exception"

File: src/arch/riscv/traps.S [hunk 2]
```diff
@@ -98,50 +98,43 @@ trap_entry:
...
+  li   s4, 8
   bne  s0, s4, exception
 
-syscall:
+  /* If it's not an exception or an interrupt it's a syscall */
```
Comment (axel-h) [inline src/arch/riscv/traps.S:108]: That comment is a bit misleading actually, better simply say "handle syscall" now.

File: src/arch/riscv/traps.S [hunk 3]
```diff
@@ -98,50 +98,43 @@ trap_entry:
...
+syscall:
+
+  /* move syscall number to 3rd argument */
   mv a2, a7
 
   j c_handle_syscall
 
 fp_reply_recv:
 
-  addi x1, x1, 4
```
Comment (axel-h) [inline src/arch/riscv/traps.S:127]: What is the commit policy actually concerning trivial changes, should removing this be a separate commit?

**Comment by axel-h** [comment]
change looks ok to me.

File: src/arch/riscv/traps.S [hunk 4]
```diff
@@ -98,50 +98,43 @@ trap_entry:
...
+#endif
+
+syscall:
+
+  /* move syscall number to 3rd argument */
   mv a2, a7
 
   j c_handle_syscall
 
 fp_reply_recv:
```
Comment (yyshen) [inline src/arch/riscv/traps.S:129]: why not combine this to the code above, just as the c_handle_fastpath_call?

File: src/arch/riscv/traps.S [hunk 5]
```diff
@@ -98,50 +98,43 @@ trap_entry:
...
-  addi x1, x1, 4
-  /* Save NextIP */
-  STORE   x1, (34*REGBYTES)(t0)
-
 #ifdef CONFIG_KERNEL_MCS
+  /* move reply to 3rd argument */
   mv a2, a6
 #endif
 
   j c_handle_fastpath_reply_recv
```
Comment (yyshen) [inline src/arch/riscv/traps.S:136]: will this cause a compiler error when FASTPATH is not enabled?

Comment (oliver-wm) [reply src/arch/riscv/traps.S:136]: Yes, you're right.

File: src/arch/riscv/traps.S (hunk 4 above)
Comment (oliver-wm) [reply src/arch/riscv/traps.S:129]: I'll do that.

**Comment by oliver-wm** [comment]
@yyshen happy with it?

**Review by xurtis — APPROVED** [review]

**Comment by yyshen** [comment]
> @yyshen happy with it?

Yes, I was about to click the approve button.

**Comment by oliver-wm** [comment]
Thanks!
